### PR TITLE
Koltok Bohdan HW 3

### PR DESCRIPTION
**Pull request by R-accoo-n**
Completed and ready to be torn apart

### CODE REVIEW

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.context.annotation.ComponentScan;
+
+/**
+ * Configuration file.
+ */
+
+@org.springframework.context.annotation.Configuration
+@ComponentScan("co.inventorsoft.academy.spring.models")
+@ComponentScan("co.inventorsoft.academy.spring.repositories")
+@ComponentScan("co.inventorsoft.academy.spring.services")
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java]: is this component scan necessary?

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.util.Set;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.ComponentScan;
+
+/**
+ * Configuration file.
+ */
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java]: the name of the class tells what it is, remove this doc

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+@ComponentScan("co.inventorsoft.academy.spring.repositories")
+@ComponentScan("co.inventorsoft.academy.spring.services")
+public class Configuration {
+    @Value("${excluded.words}")
+    private Set<String> excludedWords;
+
+    @Bean
+    public Set<String> excludedWords() {
+        return excludedWords;
+    }
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java]: it is fine but why don't directly use @Value in the class it is intended

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/models/Article.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * Article model.
+ */
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@ToString
+@Table(name = "articles")
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/models/Article.java]: it is too much, the task is not about storing data in db, remove all of these, should be just pojo

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/models/NotificationType.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package co.inventorsoft.academy.spring.models;
+
+/**
+ * Notification type enumeration model.
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/models/NotificationType.java]: The enum name tells exactly what this doc tells, remove it (also check in all other classes)

File: article-analyzer/build.gradle [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+		extendsFrom annotationProcessor
+	}
+}
+
+repositories {
+	mavenCentral()
+}
+
+dependencies {
+	implementation 'org.springframework.boot:spring-boot-starter-web'
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/build.gradle]: all we need here are spring-boot-starter, gson and lombock
postgres, jpa, h2 - no
it is nice that you know these, but we will use these things later in the appropriate lectures.
here we only focused on Spring basic features, no fancy stuff

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/models/User.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * User model.
+ */
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@ToString
+@Table(name = "users")
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/models/User.java]: leave just pojo without jpa

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/ArticleRepository.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Repository;
+
+
+/**
+ * Data access layer for Article entity.
+ */
+@Repository
+//public interface ArticleRepository extends JpaRepository<Article, UUID> {
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/ArticleRepository.java]: no Jpa repository, remove

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/ArticleRepository.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    @Value("${ARTICLES_FILE}")
+    String articlesFile;
+    private final Gson gson;
+
+    @Autowired
+    public ArticleRepository() {
+        this.gson = new GsonBuilder()
+            .registerTypeAdapter(UUID.class, new UUIDDeserializer())
+            .registerTypeAdapter(UUID.class, new UUIDSerializer())
+            .create();
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/ArticleRepository.java]: why these serializer/deserializer is needed?

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/UserRepository.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Repository;
+
+/**
+ * Data access layer for User entity.
+ */
+@Repository
+//public interface UserRepository extends JpaRepository<User, UUID> {
+public class UserRepository {
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/UserRepository.java:21]: same here, remove all jpa and uuid stuff

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/services/CategoriesService.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+    }
+
+    private void saveCategories(Set<String> categories){
+        System.out.println(categories.toString());
+        this.categoryRepository.saveCategories(categories);
+    }
+
+    public void processCategories(List<Article> articles){
+        saveCategories(processArticlesIntoCategories(articles));
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/services/CategoriesService.java]: avoid using nested method calls, move into variable

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/services/NotificationEmailService.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            .username("userNotFound")
+            .slackId("userNotFound")
+            .build()
+        );
+
+        if(currentUser.getNotificationType().equals(NotificationType.EMAIL)){
+            System.out.println("Dear " + currentUser.getUsername() + ", you have new notification:\n" + notification);
+        }else {
+            throw new UnsupportedOperationException("Wrong notification type");
+        }
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/java/co/inventorsoft/academy/spring/services/NotificationEmailService.java]: it is better just not to do anything when the wrong type

File: article-analyzer/src/main/resources/articles.json [hunk 13]
```diff
@@ -1,16 +1,16 @@
 [
   {
-    "id": 1,
+    "id": "123e4567-e89b-12d3-a456-426614174001",
```
Comment (vitaliiStefanchakInventorsoft) [inline article-analyzer/src/main/resources/articles.json]: why did you change it to UUID?

**Review by vitaliiStefanchakInventorsoft — COMMENTED** [review]
Please, strictly follow the task requirements

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/ArticleRepository.java (hunk 9 above)
Comment (R-accoo-n) [reply article-analyzer/src/main/java/co/inventorsoft/academy/spring/repositories/ArticleRepository.java]: I used UUID type and gson didnt work well with it so i used custom serializer and deserializer, but it wasn't problem with them and when i solved it completely forgot about this custom stuff, guess i should remove them

File: article-analyzer/src/main/resources/articles.json (hunk 13 above)
Comment (R-accoo-n) [reply article-analyzer/src/main/resources/articles.json]: Just tought it would be better to use uuid type for id

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java (hunk 1 above)
Comment (R-accoo-n) [reply article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java]: It was at the beginning and i forgot to remove it at the end, gonna remove it now

File: article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java (hunk 2 above)
Comment (R-accoo-n) [reply article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java]: If changed into more meaningful sentence, can it be left?

Comment (vitaliiStefanchakInventorsoft) [reply article-analyzer/src/main/java/co/inventorsoft/academy/spring/config/Configuration.java]: if so, sure

File: article-analyzer/src/main/resources/articles.json (hunk 13 above)
Comment (vitaliiStefanchakInventorsoft) [reply article-analyzer/src/main/resources/articles.json]: do not modify business data, work with what you have, in real projects that is exactly the case, there it is very seldom you have what you which btw)

**Review by vitaliiStefanchakInventorsoft — APPROVED** [review]
